### PR TITLE
Custom component gallery

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Trying something a bit different - embedding the gallery in the website! We should probably do this for the themes too - I think it will help us stay on top of it and make sure it's working.

You can test by going to the `/custom_components/gallery` route in the deployed website URL.

### My screen recorder messed up the colors!
custom_component_gallery_preview

Overview:
* There's a space serving as the backend. Right now it has a curated list of components (manually excluding test components). Going to set up the webhook next week (need help from hub team) so that the backend gets updated as components are published.
* Display all the custom component spaces in a grid kind of like we do for the guides. The grid view uses the color gradient as background like we do in the guides. Use an emoji as a background like spaces on the hub.
* I don't think we should preview the space via iframe because we got feedback that it did not scale well for the theme gallery (theme gallery was only loading like 25 spaces at a time or something).
* Instead, when you click on a space, a sidebar opens up with some info on the space and the iframe where you can view the space. We can't use the <gradio-app> embedding tag because I don't think it can load custom components?

TODO:

- [x] Fix styling
- [ ] Search bar at the top where you can search by name or tag

Out of Scope:
* sort. Not enough custom components yet 😉.
* pagination. Although the backend already supports this.

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

**Comment by abidlabs** [comment]
Took a look over this. Very cool @freddyaboulton! 

Had some questions:

> There's a space serving as the backend. Right now it has a curated list of components (manually excluding test components). Going to set up the webhook next week (need help from hub team) so that the backend gets updated as components are published.

Why are we curating a list of components instead of getting Spaces via API filtering by the `gradio-custom-component` tag (and potentially having an exclude list to exclude the test components)

On the UI side (minor points):

image

* The colors seem very faded out. I think we can make them a bit more appealing
* Font and spacing for the tags looks a bit tight

image

* I really like @pngwn's docs Space -- maybe we can just replace this preview entirely with that Space? Feels like it would be cleaner
* nit: the `gradio_pdf` Space is down

Misc:

* Nit: I was expecting the search bar to refresh the results on every keypress down like we do with the Guides

**Comment by abidlabs** [comment]
Btw I think this is fine to merge in -- I think we can do the UI improvements when @pngwn merges his doc space and then publish this webpage on the website at the point

**Comment by abidlabs** [comment]
In terms of where to put this, I'd love to get @aliabd's input, but I think the new `/docs` page can have one section for custom components, which would then pull up this page (maybe we can add some text above the search bar providing context for custom components and linking to the guides to build your own custom components)

**Comment by pngwn** [comment]
I think we can merge this in now @freddyaboulton. We can make any tweaks + surface it later. Don't want it to get stale.

File: js/_website/src/routes/custom_components/gallery/utils.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	HighlightedText: "✨",
+	HTML: "🔗",
+	Image: "🖼️",
+	JSON: "📝",
+	Label: "🏷️",
+	Markdown: "📝",
+	Model3D: "🗿",
+	State: "🔢",
+	UploadButton: "📤",
+	Video: "🎥"
```
Comment (pngwn) [inline js/_website/src/routes/custom_components/gallery/utils.ts:42]: Nice emojis.

File: js/_website/src/routes/custom_components/gallery/Close.svelte
Comment (pngwn) [inline js/_website/src/routes/custom_components/gallery/Close.svelte:1]: Lets move this to `src/lib/icons`

File: js/_website/src/routes/custom_components/gallery/CopyButton.svelte
Comment (pngwn) [inline js/_website/src/routes/custom_components/gallery/CopyButton.svelte:1]: Same with this, move to `src/lib/icons`

File: js/_website/src/routes/custom_components/gallery/+page.svelte [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+		["from-purple-100", "to-purple-50"],
+		["from-green-100", "to-green-50"],
+		["from-yellow-100", "to-yellow-50"],
+		["from-red-100", "to-red-50"],
+		["from-blue-100", "to-blue-50"],
+		["from-pink-100", "to-pink-50"],
+		["from-purple-100", "to-purple-50"]
+	];
+
+	function randomColor(): string {
```
Comment (pngwn) [inline js/_website/src/routes/custom_components/gallery/+page.svelte:32]: snake_case

File: js/_website/src/routes/custom_components/gallery/+page.svelte [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+		const color = COLOR_SETS[getRandomIntInclusive(0, COLOR_SETS.length - 1)];
+		return `${color[0]} ${color[1]}`;
+	}
+
+	const handleBoxClick = (component: ComponentData) => {
```
Comment (pngwn) [inline js/_website/src/routes/custom_components/gallery/+page.svelte:37]: snake_case

File: js/_website/src/routes/custom_components/gallery/+page.svelte [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+		selectedComponent = component;
+	};
+
+	async function fetchComponents(selection: string[] = []) {
```
Comment (pngwn) [inline js/_website/src/routes/custom_components/gallery/+page.svelte:41]: snake_case

File: js/_website/src/routes/custom_components/gallery/+page.svelte [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+	import Close from "./Close.svelte";
+
+	const API = "https://gradio-custom-component-gallery-backend.hf.space/";
+	const OFFSET = 0;
+	const LIMIT = 50;
+
+	let components: ComponentData[] = [];
+	let selection: string = "";
+
+	let selectedComponent: ComponentData | null = null;
```
Comment (pngwn) [inline js/_website/src/routes/custom_components/gallery/+page.svelte:15]: snake_case

File: js/_website/src/routes/custom_components/gallery/Card.svelte [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+	import { onMount } from "svelte";
+
+	export let data: ComponentData;
+
+	let sourceCodeLink = `https://huggingface.co/spaces/${data.id}`;
+	let authorLink = `https://huggingface.co/${data.author}`;
+	let discussionLink = `https://huggingface.co/spaces/${data.id}/discussions/new`;
+	let codeUrl = `https://huggingface.co/spaces/${data.id}/raw/main/app.py`;
+	const tabs = ["Demo", "Code"];
+	let activeTab = 0;
```
Comment (pngwn) [inline js/_website/src/routes/custom_components/gallery/Card.svelte:14]: snake_case

File: js/_website/src/routes/custom_components/gallery/utils.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	template: string;
+	author: string;
+	description: string;
+	tags: string;
+	version: string;
+	subdomain: string;
+	background_color: string;
+};
+
+export function getRandomIntInclusive(min: number, max: number): number {
```
Comment (pngwn) [inline js/_website/src/routes/custom_components/gallery/utils.ts:13]: snake_case

File: js/_website/src/routes/custom_components/gallery/utils.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	min = Math.ceil(min);
+	max = Math.floor(max);
+	return Math.floor(Math.random() * (max - min + 1) + min); // The maximum is inclusive and the minimum is inclusive
+}
+
+export const classToEmojiMapping: { [key: string]: string } = {
```
Comment (pngwn) [inline js/_website/src/routes/custom_components/gallery/utils.ts:19]: snake_case

**Review by pngwn — APPROVED** [review]
Just some casing changes for consistency. Looks great otherwise. 

I think we can merge this now. We can figure out any tweaks and how/ where to surface it in a later PR. Don't want it to get stale!

Thanks for tackling this @freddyaboulton, it will be a huge help with custom components!

**Review by aliabd — APPROVED** [review]
Agree with @pngwn, let's merge this for now and I can find a home for it in the website in a different PR. Like @abidlabs said, will be at /docs/custom-components along with the cc guides (once I'm done with docs reorg...) 

Nice work @freddyaboulton!

**Comment by freddyaboulton** [comment]
Thanks everyone for the reviews! Can't wait to surface in docs once the re-org is done and to use the auto doc component in the preview

**Pull request merged**